### PR TITLE
Assure address of PC at startup.

### PR DESCRIPTION
**Pull request by elfmimi**
PR created upon @Disasm 's request. 

Closes: #44

Need update of library files after merging.

### CODE REVIEW

**Comment by almindor** [comment]
Seems the expected elf/linux binaries need fixing for CI

File: asm.S [hunk 1]
```diff
@@ -12,6 +12,25 @@
...
+    .if __riscv_xlen == 32
+    lui ra, %hi(_abs_start)
+    jr %lo(_abs_start)(ra)
+    .endif
+
+    // for 64bit
+    .if __riscv_xlen == 64
+1:
+    auipc ra, %pcrel_hi(1f)
+    ld ra, %pcrel_lo(1b)(ra)
```
Comment (Disasm) [inline asm.S]: Should this be `%pcrel_lo(1f)(ra)`?

**Comment by Disasm** [comment]
@elfmimi Thank you! Could you run `assemble.sh` and commit the changed `bin` directory? You can use the toolchain from SiFive for this.

File: asm.S (hunk 1 above)
Comment (elfmimi) [reply asm.S]: Here is a documentation on it.
https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#pc-relative-symbol-addresses

Comment (Disasm) [reply asm.S]: Indeed, thank you for pointing out!

**Comment by piegamesde** [comment]
This change gives me headaches. Can somebody please explain to somebody less familiar with RISC-V assembly what this code exactly does, and why? For example, why isn't the 32 bit code not a simple `j _abs_start`? And why is the 64 bit version so much more complicated?

**Comment by Disasm** [comment]
All the binaries linked for embedded targets have "assumed" load address and they are linked for this static address (and there is no dynamic relocation support and pc-relative addressing). However, the reality is sometimes different. For example, K210 has two regions or RAM: cached and non-cached. They are mapped to the same RAM, but non-cached region is slow. Bootloader jumps into the non-cached region, so we need to jump back to the cached region to achieve expected performance. Similar thing happens on GD32V: it has two regions (which have the same properties this time): starting from 0 and 0x0800_0000. CPU core jumps into 0, but we expect 0x0800_0000. Of course, we could modify linker script to link everything into 0, but then we will have problems with flashing the binary into the microcontroller.

With `j _abs_start` you can't jump to any 32-bit address (the instruction itself is 32-bit so you can't encode all the address bits), so you have to load the address separately. If you jump to `_abs_start` with relative-address jump, you will end up in the same memory region because `_abs_start` is close to the jumping point, so you have to use absolute address jumps.
Then, on 32-systems you can load any absolute 32-bit address with two instructions, which is not the case for 64-bit systems because the address is twice larger. So for 64-bit targets we have to store the address into a global variable and then load it using pc-relative addressing.

**Pull request merged**